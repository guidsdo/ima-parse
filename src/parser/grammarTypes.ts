import { CharCode } from "../helpers/charCodeHelpers";

type DefinitionPartBase = {
    type: string;
    key?: string;
    description?: string;
    optional?: boolean;
};

/**
 * Basically // comments and "texts"
 */
export type DefinitionText = DefinitionPartBase & {
    type: "text";
    key: string;
    startPhrase: string;
    endPhrase: string;
};

/**
 * Modifiers where the order doesn't matter. If the order matters, use separate definitions.
 */
export type DefinitionModifiers = DefinitionPartBase & {
    type: "modifiers";
    phrases: string[];
    singular?: boolean;
    key?: never;
};

/**
 * Alternative paths that can only contain simple definition parts. This is to allow more flexibility but still restrict the complexity.
 * The parts of all the paths will be added to the typings of the containing rule with optionality.
 */
export type DefinitionPaths = DefinitionPartBase & {
    type: "paths";
    paths: SimpleDefinitionPart[][];
    key?: never;
};

/**
 * A required keyword/character that provides structure to the rule that's being defined.
 * Optional? No, always required.
 */
export type DefinitionKeyword = DefinitionPartBase & {
    type: "keyword";
    phrase: string;
    optional?: false;
    key?: never;
};

/**
 * A number definition. Need other types of characters in front or behind? Use paths and/or modifiers.
 * Optional? No, always required.
 */
export type DefinitionNumber = DefinitionPartBase & {
    type: "number";
    key: string;
    optional?: false;
};

/**
 * Identifier that can be a word that contains letters, numbers and underscores.
 * Optional? No, always required.
 */
export type DefinitionIdentifier = DefinitionPartBase & {
    type: "identifier";
    key: string;
    optional?: false;
};

/**
 * Rules that always are a child of another rule.
 */
export type DefinitionRules = DefinitionPartBase & {
    type: "rules";
    key: string;
    rules: GrammarRule[] | (() => GrammarRule[]);
    optional: boolean;
    /** Only allow one of the rules once */
    singular?: boolean;
    /** Phrase that separates the rules */
    separatorPhrase?: string;
    /* Unfortunately this is something encountered */
    separatorOptional?: boolean;
};

export type SimpleDefinitionPart = DefinitionIdentifier | DefinitionModifiers | DefinitionKeyword | DefinitionText | DefinitionNumber;

export type DefinitionPart = SimpleDefinitionPart | DefinitionRules | DefinitionPaths;

/**
 * Define a 'rule' that can be referred to from other rules (where you allow it). Make sure that the definition parts within are
 * NON-ambiguous! All array definitions are parsed in order, so take that into account when defining.
 */
export type GrammarRule = { name: string; definition: DefinitionPart[] };

export type Grammar = {
    TopLevel: GrammarRule;
    global: DefinitionRules;
    wordChars?: CharCode[];
    numberChars?: CharCode[];
    validChars?: CharCode[];
};
