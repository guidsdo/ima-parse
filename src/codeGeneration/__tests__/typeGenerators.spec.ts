import ts from "typescript";
import { Parser } from "../../parser/Parser";
import { Grammar } from "../../parser/grammarTypes";
import { classRule, importRule } from "../../parser/__tests__/testHelpers";
import { getTypesAndBuildersFromGrammar, buildContentTree } from "../typeGenerators";

describe("typeGenerators", () => {
    describe("getTypesAndBuildersFromGrammar", () => {
        it("should generate the types for the given grammar", () => {
            const grammar: Grammar = {
                TopLevel: {
                    name: "TopLevel",
                    definition: [{ type: "rules", optional: true, key: "content", rules: [importRule, classRule] }]
                },
                global: { type: "rules", optional: true, key: "content", rules: [] }
            };

            const { types } = getTypesAndBuildersFromGrammar(grammar);

            expect(`\n${types.join("\n\n")}`).toStrictEqual(
                `
export type $Cursor = { ln: number; col: number };

export type $Position = { start: $Cursor; end: $Cursor };

export type $RuleContentTreeBase = { $type: string; $position: $Position; }

export type $RuleContentTree = TopLevel | Import | ClassDefinition | ImportParts | Asterisk | StringLiteral | ClassProperty | Identifier | ValueAssignment | NumberLiteral

export type TopLevel = $RuleContentTreeBase & {
    $type: "TopLevel";
    "content"?: { type: "children"; value: (Import | ClassDefinition)[]; position: $Position; };
};

export type Import = $RuleContentTreeBase & {
    $type: "Import";
    $keywords: { type: "keyword"; value: string; position: $Position; }[];
    "target": { type: "child"; value: ImportParts | Asterisk; position: $Position; };
    "source": { type: "child"; value: StringLiteral; position: $Position; };
};

export type ClassDefinition = $RuleContentTreeBase & {
    $type: "ClassDefinition";
    $keywords: { type: "keyword"; value: string; position: $Position; }[];
    "abstract"?: { type: "modifier"; value: boolean; position: $Position; };
    "name": { type: "identifier"; value: string; position: $Position; };
    "extends"?: { type: "identifier"; value: string; position: $Position; };
    "properties"?: { type: "children"; value: (ClassProperty)[]; position: $Position; };
};

export type ImportParts = $RuleContentTreeBase & {
    $type: "ImportParts";
    $keywords: { type: "keyword"; value: string; position: $Position; }[];
    "values": { type: "children"; value: (Identifier)[]; position: $Position; };
};

export type Asterisk = $RuleContentTreeBase & {
    $type: "Asterisk";
    $keywords: { type: "keyword"; value: string; position: $Position; }[];
};

export type StringLiteral = $RuleContentTreeBase & {
    $type: "StringLiteral";
    "value": { type: "text"; value: string; position: $Position; };
};

export type ClassProperty = $RuleContentTreeBase & {
    $type: "ClassProperty";
    $keywords?: { type: "keyword"; value: string; position: $Position; }[];
    "name": { type: "identifier"; value: string; position: $Position; };
    "type"?: { type: "identifier"; value: string; position: $Position; };
    "defaultValue"?: { type: "child"; value: ValueAssignment; position: $Position; };
};

export type Identifier = $RuleContentTreeBase & {
    $type: "Identifier";
    "name": { type: "identifier"; value: string; position: $Position; };
};

export type ValueAssignment = $RuleContentTreeBase & {
    $type: "ValueAssignment";
    $keywords: { type: "keyword"; value: string; position: $Position; }[];
    "value": { type: "child"; value: StringLiteral | NumberLiteral; position: $Position; };
};

export type NumberLiteral = $RuleContentTreeBase & {
    $type: "NumberLiteral";
    "number": { type: "number"; value: number; position: $Position; };
};`
            );
        });
    });

    describe("buildContentTree", () => {
        it("should fill the content tree with the parsed text and adhere to the generated types", () => {
            const grammar: Grammar = {
                TopLevel: { name: "T", definition: [{ type: "rules", optional: true, key: "content", rules: [classRule, importRule] }] },
                global: { type: "rules", optional: true, key: "content", rules: [] }
            };

            const input = `
        import {k, bla } from "./someLibrary"
        import * from "path"

        abstract class Animal {
            legs: number = 4
        }

        abstract class Beast extends Animal{
            legs: number = 393483493489
        }

        class	dog	extends	Animal		{
            otherProperty
            AnotherProperty: string
            okOneMore = "lol"
            forReal:string="yes"
        }

     import {bonus}from"lib"
        `;

            // Parse the text
            const parser = new Parser(grammar);
            parser.parseText(input);
            expect(parser.brokenContent.length).toStrictEqual(0);

            const { types, ruleBuilders } = getTypesAndBuildersFromGrammar(grammar);

            // Act
            const builtTree = buildContentTree(ruleBuilders, parser.getTopLevelParser());

            // Assert: Have a snapshot to always validate that there is a real tree being generated
            expect(builtTree).toMatchSnapshot();

            // Assert: Check if this code adheres to the generated types by letting Typescript compile it
            const typesContent = types.join("\n\n");
            const valueContent = JSON.stringify(builtTree, null, "  ");
            const code = `${typesContent}\n\nconst x: T = ${valueContent}`;

            const tsDiagnostics = compileTypeScriptCode(code);

            expect(tsDiagnostics.length).toStrictEqual(0);
        });
    });
});

function compileTypeScriptCode(code: string, libs: string[] = ["es2022"]): ts.Diagnostic[] {
    const options = ts.getDefaultCompilerOptions();
    const realHost = ts.createCompilerHost(options, true);

    const dummyFilePath = "/in-memory-file.ts";
    const dummySourceFile = ts.createSourceFile(dummyFilePath, code, ts.ScriptTarget.Latest);

    let outputCode: string | undefined = undefined;

    const host: ts.CompilerHost = {
        fileExists: filePath => filePath === dummyFilePath || realHost.fileExists(filePath),
        directoryExists: realHost.directoryExists && realHost.directoryExists.bind(realHost),
        getCurrentDirectory: realHost.getCurrentDirectory.bind(realHost),
        getDirectories: realHost.getDirectories?.bind(realHost),
        getCanonicalFileName: fileName => realHost.getCanonicalFileName(fileName),
        getNewLine: realHost.getNewLine.bind(realHost),
        getDefaultLibFileName: realHost.getDefaultLibFileName.bind(realHost),
        getSourceFile: (fileName, languageVersion, onError, shouldCreateNewSourceFile) =>
            fileName === dummyFilePath
                ? dummySourceFile
                : realHost.getSourceFile(fileName, languageVersion, onError, shouldCreateNewSourceFile),
        readFile: filePath => (filePath === dummyFilePath ? code : realHost.readFile(filePath)),
        useCaseSensitiveFileNames: () => realHost.useCaseSensitiveFileNames(),
        writeFile: (_, data) => (outputCode = data)
    };

    const rootNames = libs.map(lib => require.resolve(`typescript/lib/lib.${lib}.d.ts`));
    const program = ts.createProgram(rootNames.concat([dummyFilePath]), options, host);
    const emitResult = program.emit();
    const diagnostics = ts.getPreEmitDiagnostics(program);
    return emitResult.diagnostics.concat(diagnostics);
}
