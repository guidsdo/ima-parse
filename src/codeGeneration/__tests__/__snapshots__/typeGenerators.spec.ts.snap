// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`typeGenerators buildContentTree should fill the content tree with the parsed text and adhere to the generated types 1`] = `
{
  "$position": {
    "end": {
      "col": 25,
      "ln": 20,
    },
    "start": {
      "col": 9,
      "ln": 2,
    },
  },
  "$type": "T",
  "content": {
    "position": {
      "end": {
        "col": 25,
        "ln": 20,
      },
      "start": {
        "col": 9,
        "ln": 2,
      },
    },
    "type": "children",
    "value": [
      {
        "$keywords": [
          {
            "position": {
              "end": {
                "col": 15,
                "ln": 2,
              },
              "start": {
                "col": 9,
                "ln": 2,
              },
            },
            "type": "keyword",
            "value": "import",
          },
          {
            "position": {
              "end": {
                "col": 30,
                "ln": 2,
              },
              "start": {
                "col": 26,
                "ln": 2,
              },
            },
            "type": "keyword",
            "value": "from",
          },
        ],
        "$position": {
          "end": {
            "col": 32,
            "ln": 2,
          },
          "start": {
            "col": 9,
            "ln": 2,
          },
        },
        "$type": "Import",
        "source": {
          "position": {
            "end": {
              "col": 32,
              "ln": 2,
            },
            "start": {
              "col": 31,
              "ln": 2,
            },
          },
          "type": "child",
          "value": {
            "$position": {
              "end": {
                "col": 32,
                "ln": 2,
              },
              "start": {
                "col": 31,
                "ln": 2,
              },
            },
            "$type": "StringLiteral",
            "value": {
              "position": {
                "end": {
                  "col": 32,
                  "ln": 2,
                },
                "start": {
                  "col": 31,
                  "ln": 2,
                },
              },
              "type": "text",
              "value": ""./someLibrary"",
            },
          },
        },
        "target": {
          "position": {
            "end": {
              "col": 25,
              "ln": 2,
            },
            "start": {
              "col": 16,
              "ln": 2,
            },
          },
          "type": "child",
          "value": {
            "$keywords": [
              {
                "position": {
                  "end": {
                    "col": 17,
                    "ln": 2,
                  },
                  "start": {
                    "col": 16,
                    "ln": 2,
                  },
                },
                "type": "keyword",
                "value": "{",
              },
              {
                "position": {
                  "end": {
                    "col": 25,
                    "ln": 2,
                  },
                  "start": {
                    "col": 24,
                    "ln": 2,
                  },
                },
                "type": "keyword",
                "value": "}",
              },
            ],
            "$position": {
              "end": {
                "col": 25,
                "ln": 2,
              },
              "start": {
                "col": 16,
                "ln": 2,
              },
            },
            "$type": "ImportParts",
            "values": {
              "position": {
                "end": {
                  "col": 23,
                  "ln": 2,
                },
                "start": {
                  "col": 17,
                  "ln": 2,
                },
              },
              "type": "children",
              "value": [
                {
                  "$position": {
                    "end": {
                      "col": 18,
                      "ln": 2,
                    },
                    "start": {
                      "col": 17,
                      "ln": 2,
                    },
                  },
                  "$type": "Identifier",
                  "name": {
                    "position": {
                      "end": {
                        "col": 18,
                        "ln": 2,
                      },
                      "start": {
                        "col": 17,
                        "ln": 2,
                      },
                    },
                    "type": "identifier",
                    "value": "k",
                  },
                },
                {
                  "$position": {
                    "end": {
                      "col": 23,
                      "ln": 2,
                    },
                    "start": {
                      "col": 20,
                      "ln": 2,
                    },
                  },
                  "$type": "Identifier",
                  "name": {
                    "position": {
                      "end": {
                        "col": 23,
                        "ln": 2,
                      },
                      "start": {
                        "col": 20,
                        "ln": 2,
                      },
                    },
                    "type": "identifier",
                    "value": "bla",
                  },
                },
              ],
            },
          },
        },
      },
      {
        "$keywords": [
          {
            "position": {
              "end": {
                "col": 15,
                "ln": 3,
              },
              "start": {
                "col": 9,
                "ln": 3,
              },
            },
            "type": "keyword",
            "value": "import",
          },
          {
            "position": {
              "end": {
                "col": 22,
                "ln": 3,
              },
              "start": {
                "col": 18,
                "ln": 3,
              },
            },
            "type": "keyword",
            "value": "from",
          },
        ],
        "$position": {
          "end": {
            "col": 24,
            "ln": 3,
          },
          "start": {
            "col": 9,
            "ln": 3,
          },
        },
        "$type": "Import",
        "source": {
          "position": {
            "end": {
              "col": 24,
              "ln": 3,
            },
            "start": {
              "col": 23,
              "ln": 3,
            },
          },
          "type": "child",
          "value": {
            "$position": {
              "end": {
                "col": 24,
                "ln": 3,
              },
              "start": {
                "col": 23,
                "ln": 3,
              },
            },
            "$type": "StringLiteral",
            "value": {
              "position": {
                "end": {
                  "col": 24,
                  "ln": 3,
                },
                "start": {
                  "col": 23,
                  "ln": 3,
                },
              },
              "type": "text",
              "value": ""path"",
            },
          },
        },
        "target": {
          "position": {
            "end": {
              "col": 17,
              "ln": 3,
            },
            "start": {
              "col": 16,
              "ln": 3,
            },
          },
          "type": "child",
          "value": {
            "$keywords": [
              {
                "position": {
                  "end": {
                    "col": 17,
                    "ln": 3,
                  },
                  "start": {
                    "col": 16,
                    "ln": 3,
                  },
                },
                "type": "keyword",
                "value": "*",
              },
            ],
            "$position": {
              "end": {
                "col": 17,
                "ln": 3,
              },
              "start": {
                "col": 16,
                "ln": 3,
              },
            },
            "$type": "Asterisk",
          },
        },
      },
      {
        "$keywords": [
          {
            "position": {
              "end": {
                "col": 23,
                "ln": 5,
              },
              "start": {
                "col": 18,
                "ln": 5,
              },
            },
            "type": "keyword",
            "value": "class",
          },
          {
            "position": {
              "end": {
                "col": 32,
                "ln": 5,
              },
              "start": {
                "col": 31,
                "ln": 5,
              },
            },
            "type": "keyword",
            "value": "{",
          },
          {
            "position": {
              "end": {
                "col": 10,
                "ln": 7,
              },
              "start": {
                "col": 9,
                "ln": 7,
              },
            },
            "type": "keyword",
            "value": "}",
          },
        ],
        "$position": {
          "end": {
            "col": 10,
            "ln": 7,
          },
          "start": {
            "col": 9,
            "ln": 5,
          },
        },
        "$type": "ClassDefinition",
        "abstract": {
          "position": {
            "end": {
              "col": 17,
              "ln": 5,
            },
            "start": {
              "col": 9,
              "ln": 5,
            },
          },
          "type": "modifier",
          "value": true,
        },
        "name": {
          "position": {
            "end": {
              "col": 30,
              "ln": 5,
            },
            "start": {
              "col": 24,
              "ln": 5,
            },
          },
          "type": "identifier",
          "value": "Animal",
        },
        "properties": {
          "position": {
            "end": {
              "col": 29,
              "ln": 6,
            },
            "start": {
              "col": 13,
              "ln": 6,
            },
          },
          "type": "children",
          "value": [
            {
              "$keywords": [
                {
                  "position": {
                    "end": {
                      "col": 18,
                      "ln": 6,
                    },
                    "start": {
                      "col": 17,
                      "ln": 6,
                    },
                  },
                  "type": "keyword",
                  "value": ":",
                },
              ],
              "$position": {
                "end": {
                  "col": 29,
                  "ln": 6,
                },
                "start": {
                  "col": 13,
                  "ln": 6,
                },
              },
              "$type": "ClassProperty",
              "defaultValue": {
                "position": {
                  "end": {
                    "col": 29,
                    "ln": 6,
                  },
                  "start": {
                    "col": 26,
                    "ln": 6,
                  },
                },
                "type": "child",
                "value": {
                  "$keywords": [
                    {
                      "position": {
                        "end": {
                          "col": 27,
                          "ln": 6,
                        },
                        "start": {
                          "col": 26,
                          "ln": 6,
                        },
                      },
                      "type": "keyword",
                      "value": "=",
                    },
                  ],
                  "$position": {
                    "end": {
                      "col": 29,
                      "ln": 6,
                    },
                    "start": {
                      "col": 26,
                      "ln": 6,
                    },
                  },
                  "$type": "ValueAssignment",
                  "value": {
                    "position": {
                      "end": {
                        "col": 29,
                        "ln": 6,
                      },
                      "start": {
                        "col": 28,
                        "ln": 6,
                      },
                    },
                    "type": "child",
                    "value": {
                      "$position": {
                        "end": {
                          "col": 29,
                          "ln": 6,
                        },
                        "start": {
                          "col": 28,
                          "ln": 6,
                        },
                      },
                      "$type": "NumberLiteral",
                      "number": {
                        "position": {
                          "end": {
                            "col": 29,
                            "ln": 6,
                          },
                          "start": {
                            "col": 28,
                            "ln": 6,
                          },
                        },
                        "type": "number",
                        "value": 4,
                      },
                    },
                  },
                },
              },
              "name": {
                "position": {
                  "end": {
                    "col": 17,
                    "ln": 6,
                  },
                  "start": {
                    "col": 13,
                    "ln": 6,
                  },
                },
                "type": "identifier",
                "value": "legs",
              },
              "type": {
                "position": {
                  "end": {
                    "col": 25,
                    "ln": 6,
                  },
                  "start": {
                    "col": 19,
                    "ln": 6,
                  },
                },
                "type": "identifier",
                "value": "number",
              },
            },
          ],
        },
      },
      {
        "$keywords": [
          {
            "position": {
              "end": {
                "col": 23,
                "ln": 9,
              },
              "start": {
                "col": 18,
                "ln": 9,
              },
            },
            "type": "keyword",
            "value": "class",
          },
          {
            "position": {
              "end": {
                "col": 37,
                "ln": 9,
              },
              "start": {
                "col": 30,
                "ln": 9,
              },
            },
            "type": "keyword",
            "value": "extends",
          },
          {
            "position": {
              "end": {
                "col": 45,
                "ln": 9,
              },
              "start": {
                "col": 44,
                "ln": 9,
              },
            },
            "type": "keyword",
            "value": "{",
          },
          {
            "position": {
              "end": {
                "col": 10,
                "ln": 11,
              },
              "start": {
                "col": 9,
                "ln": 11,
              },
            },
            "type": "keyword",
            "value": "}",
          },
        ],
        "$position": {
          "end": {
            "col": 10,
            "ln": 11,
          },
          "start": {
            "col": 9,
            "ln": 9,
          },
        },
        "$type": "ClassDefinition",
        "abstract": {
          "position": {
            "end": {
              "col": 17,
              "ln": 9,
            },
            "start": {
              "col": 9,
              "ln": 9,
            },
          },
          "type": "modifier",
          "value": true,
        },
        "extends": {
          "position": {
            "end": {
              "col": 44,
              "ln": 9,
            },
            "start": {
              "col": 38,
              "ln": 9,
            },
          },
          "type": "identifier",
          "value": "Animal",
        },
        "name": {
          "position": {
            "end": {
              "col": 29,
              "ln": 9,
            },
            "start": {
              "col": 24,
              "ln": 9,
            },
          },
          "type": "identifier",
          "value": "Beast",
        },
        "properties": {
          "position": {
            "end": {
              "col": 40,
              "ln": 10,
            },
            "start": {
              "col": 13,
              "ln": 10,
            },
          },
          "type": "children",
          "value": [
            {
              "$keywords": [
                {
                  "position": {
                    "end": {
                      "col": 18,
                      "ln": 10,
                    },
                    "start": {
                      "col": 17,
                      "ln": 10,
                    },
                  },
                  "type": "keyword",
                  "value": ":",
                },
              ],
              "$position": {
                "end": {
                  "col": 40,
                  "ln": 10,
                },
                "start": {
                  "col": 13,
                  "ln": 10,
                },
              },
              "$type": "ClassProperty",
              "defaultValue": {
                "position": {
                  "end": {
                    "col": 40,
                    "ln": 10,
                  },
                  "start": {
                    "col": 26,
                    "ln": 10,
                  },
                },
                "type": "child",
                "value": {
                  "$keywords": [
                    {
                      "position": {
                        "end": {
                          "col": 27,
                          "ln": 10,
                        },
                        "start": {
                          "col": 26,
                          "ln": 10,
                        },
                      },
                      "type": "keyword",
                      "value": "=",
                    },
                  ],
                  "$position": {
                    "end": {
                      "col": 40,
                      "ln": 10,
                    },
                    "start": {
                      "col": 26,
                      "ln": 10,
                    },
                  },
                  "$type": "ValueAssignment",
                  "value": {
                    "position": {
                      "end": {
                        "col": 40,
                        "ln": 10,
                      },
                      "start": {
                        "col": 28,
                        "ln": 10,
                      },
                    },
                    "type": "child",
                    "value": {
                      "$position": {
                        "end": {
                          "col": 40,
                          "ln": 10,
                        },
                        "start": {
                          "col": 28,
                          "ln": 10,
                        },
                      },
                      "$type": "NumberLiteral",
                      "number": {
                        "position": {
                          "end": {
                            "col": 40,
                            "ln": 10,
                          },
                          "start": {
                            "col": 28,
                            "ln": 10,
                          },
                        },
                        "type": "number",
                        "value": 393483493489,
                      },
                    },
                  },
                },
              },
              "name": {
                "position": {
                  "end": {
                    "col": 17,
                    "ln": 10,
                  },
                  "start": {
                    "col": 13,
                    "ln": 10,
                  },
                },
                "type": "identifier",
                "value": "legs",
              },
              "type": {
                "position": {
                  "end": {
                    "col": 25,
                    "ln": 10,
                  },
                  "start": {
                    "col": 19,
                    "ln": 10,
                  },
                },
                "type": "identifier",
                "value": "number",
              },
            },
          ],
        },
      },
      {
        "$keywords": [
          {
            "position": {
              "end": {
                "col": 14,
                "ln": 13,
              },
              "start": {
                "col": 9,
                "ln": 13,
              },
            },
            "type": "keyword",
            "value": "class",
          },
          {
            "position": {
              "end": {
                "col": 26,
                "ln": 13,
              },
              "start": {
                "col": 19,
                "ln": 13,
              },
            },
            "type": "keyword",
            "value": "extends",
          },
          {
            "position": {
              "end": {
                "col": 36,
                "ln": 13,
              },
              "start": {
                "col": 35,
                "ln": 13,
              },
            },
            "type": "keyword",
            "value": "{",
          },
          {
            "position": {
              "end": {
                "col": 10,
                "ln": 18,
              },
              "start": {
                "col": 9,
                "ln": 18,
              },
            },
            "type": "keyword",
            "value": "}",
          },
        ],
        "$position": {
          "end": {
            "col": 10,
            "ln": 18,
          },
          "start": {
            "col": 9,
            "ln": 13,
          },
        },
        "$type": "ClassDefinition",
        "extends": {
          "position": {
            "end": {
              "col": 33,
              "ln": 13,
            },
            "start": {
              "col": 27,
              "ln": 13,
            },
          },
          "type": "identifier",
          "value": "Animal",
        },
        "name": {
          "position": {
            "end": {
              "col": 18,
              "ln": 13,
            },
            "start": {
              "col": 15,
              "ln": 13,
            },
          },
          "type": "identifier",
          "value": "dog",
        },
        "properties": {
          "position": {
            "end": {
              "col": 29,
              "ln": 17,
            },
            "start": {
              "col": 13,
              "ln": 14,
            },
          },
          "type": "children",
          "value": [
            {
              "$position": {
                "end": {
                  "col": 26,
                  "ln": 14,
                },
                "start": {
                  "col": 13,
                  "ln": 14,
                },
              },
              "$type": "ClassProperty",
              "name": {
                "position": {
                  "end": {
                    "col": 26,
                    "ln": 14,
                  },
                  "start": {
                    "col": 13,
                    "ln": 14,
                  },
                },
                "type": "identifier",
                "value": "otherProperty",
              },
            },
            {
              "$keywords": [
                {
                  "position": {
                    "end": {
                      "col": 29,
                      "ln": 15,
                    },
                    "start": {
                      "col": 28,
                      "ln": 15,
                    },
                  },
                  "type": "keyword",
                  "value": ":",
                },
              ],
              "$position": {
                "end": {
                  "col": 36,
                  "ln": 15,
                },
                "start": {
                  "col": 13,
                  "ln": 15,
                },
              },
              "$type": "ClassProperty",
              "name": {
                "position": {
                  "end": {
                    "col": 28,
                    "ln": 15,
                  },
                  "start": {
                    "col": 13,
                    "ln": 15,
                  },
                },
                "type": "identifier",
                "value": "AnotherProperty",
              },
              "type": {
                "position": {
                  "end": {
                    "col": 36,
                    "ln": 15,
                  },
                  "start": {
                    "col": 30,
                    "ln": 15,
                  },
                },
                "type": "identifier",
                "value": "string",
              },
            },
            {
              "$position": {
                "end": {
                  "col": 26,
                  "ln": 16,
                },
                "start": {
                  "col": 13,
                  "ln": 16,
                },
              },
              "$type": "ClassProperty",
              "defaultValue": {
                "position": {
                  "end": {
                    "col": 26,
                    "ln": 16,
                  },
                  "start": {
                    "col": 23,
                    "ln": 16,
                  },
                },
                "type": "child",
                "value": {
                  "$keywords": [
                    {
                      "position": {
                        "end": {
                          "col": 24,
                          "ln": 16,
                        },
                        "start": {
                          "col": 23,
                          "ln": 16,
                        },
                      },
                      "type": "keyword",
                      "value": "=",
                    },
                  ],
                  "$position": {
                    "end": {
                      "col": 26,
                      "ln": 16,
                    },
                    "start": {
                      "col": 23,
                      "ln": 16,
                    },
                  },
                  "$type": "ValueAssignment",
                  "value": {
                    "position": {
                      "end": {
                        "col": 26,
                        "ln": 16,
                      },
                      "start": {
                        "col": 25,
                        "ln": 16,
                      },
                    },
                    "type": "child",
                    "value": {
                      "$position": {
                        "end": {
                          "col": 26,
                          "ln": 16,
                        },
                        "start": {
                          "col": 25,
                          "ln": 16,
                        },
                      },
                      "$type": "StringLiteral",
                      "value": {
                        "position": {
                          "end": {
                            "col": 26,
                            "ln": 16,
                          },
                          "start": {
                            "col": 25,
                            "ln": 16,
                          },
                        },
                        "type": "text",
                        "value": ""lol"",
                      },
                    },
                  },
                },
              },
              "name": {
                "position": {
                  "end": {
                    "col": 22,
                    "ln": 16,
                  },
                  "start": {
                    "col": 13,
                    "ln": 16,
                  },
                },
                "type": "identifier",
                "value": "okOneMore",
              },
            },
            {
              "$keywords": [
                {
                  "position": {
                    "end": {
                      "col": 21,
                      "ln": 17,
                    },
                    "start": {
                      "col": 20,
                      "ln": 17,
                    },
                  },
                  "type": "keyword",
                  "value": ":",
                },
              ],
              "$position": {
                "end": {
                  "col": 29,
                  "ln": 17,
                },
                "start": {
                  "col": 13,
                  "ln": 17,
                },
              },
              "$type": "ClassProperty",
              "defaultValue": {
                "position": {
                  "end": {
                    "col": 29,
                    "ln": 17,
                  },
                  "start": {
                    "col": 27,
                    "ln": 17,
                  },
                },
                "type": "child",
                "value": {
                  "$keywords": [
                    {
                      "position": {
                        "end": {
                          "col": 28,
                          "ln": 17,
                        },
                        "start": {
                          "col": 27,
                          "ln": 17,
                        },
                      },
                      "type": "keyword",
                      "value": "=",
                    },
                  ],
                  "$position": {
                    "end": {
                      "col": 29,
                      "ln": 17,
                    },
                    "start": {
                      "col": 27,
                      "ln": 17,
                    },
                  },
                  "$type": "ValueAssignment",
                  "value": {
                    "position": {
                      "end": {
                        "col": 29,
                        "ln": 17,
                      },
                      "start": {
                        "col": 28,
                        "ln": 17,
                      },
                    },
                    "type": "child",
                    "value": {
                      "$position": {
                        "end": {
                          "col": 29,
                          "ln": 17,
                        },
                        "start": {
                          "col": 28,
                          "ln": 17,
                        },
                      },
                      "$type": "StringLiteral",
                      "value": {
                        "position": {
                          "end": {
                            "col": 29,
                            "ln": 17,
                          },
                          "start": {
                            "col": 28,
                            "ln": 17,
                          },
                        },
                        "type": "text",
                        "value": ""yes"",
                      },
                    },
                  },
                },
              },
              "name": {
                "position": {
                  "end": {
                    "col": 20,
                    "ln": 17,
                  },
                  "start": {
                    "col": 13,
                    "ln": 17,
                  },
                },
                "type": "identifier",
                "value": "forReal",
              },
              "type": {
                "position": {
                  "end": {
                    "col": 27,
                    "ln": 17,
                  },
                  "start": {
                    "col": 21,
                    "ln": 17,
                  },
                },
                "type": "identifier",
                "value": "string",
              },
            },
          ],
        },
      },
      {
        "$keywords": [
          {
            "position": {
              "end": {
                "col": 12,
                "ln": 20,
              },
              "start": {
                "col": 6,
                "ln": 20,
              },
            },
            "type": "keyword",
            "value": "import",
          },
          {
            "position": {
              "end": {
                "col": 24,
                "ln": 20,
              },
              "start": {
                "col": 20,
                "ln": 20,
              },
            },
            "type": "keyword",
            "value": "from",
          },
        ],
        "$position": {
          "end": {
            "col": 25,
            "ln": 20,
          },
          "start": {
            "col": 6,
            "ln": 20,
          },
        },
        "$type": "Import",
        "source": {
          "position": {
            "end": {
              "col": 25,
              "ln": 20,
            },
            "start": {
              "col": 24,
              "ln": 20,
            },
          },
          "type": "child",
          "value": {
            "$position": {
              "end": {
                "col": 25,
                "ln": 20,
              },
              "start": {
                "col": 24,
                "ln": 20,
              },
            },
            "$type": "StringLiteral",
            "value": {
              "position": {
                "end": {
                  "col": 25,
                  "ln": 20,
                },
                "start": {
                  "col": 24,
                  "ln": 20,
                },
              },
              "type": "text",
              "value": ""lib"",
            },
          },
        },
        "target": {
          "position": {
            "end": {
              "col": 20,
              "ln": 20,
            },
            "start": {
              "col": 13,
              "ln": 20,
            },
          },
          "type": "child",
          "value": {
            "$keywords": [
              {
                "position": {
                  "end": {
                    "col": 14,
                    "ln": 20,
                  },
                  "start": {
                    "col": 13,
                    "ln": 20,
                  },
                },
                "type": "keyword",
                "value": "{",
              },
              {
                "position": {
                  "end": {
                    "col": 20,
                    "ln": 20,
                  },
                  "start": {
                    "col": 19,
                    "ln": 20,
                  },
                },
                "type": "keyword",
                "value": "}",
              },
            ],
            "$position": {
              "end": {
                "col": 20,
                "ln": 20,
              },
              "start": {
                "col": 13,
                "ln": 20,
              },
            },
            "$type": "ImportParts",
            "values": {
              "position": {
                "end": {
                  "col": 19,
                  "ln": 20,
                },
                "start": {
                  "col": 14,
                  "ln": 20,
                },
              },
              "type": "children",
              "value": [
                {
                  "$position": {
                    "end": {
                      "col": 19,
                      "ln": 20,
                    },
                    "start": {
                      "col": 14,
                      "ln": 20,
                    },
                  },
                  "$type": "Identifier",
                  "name": {
                    "position": {
                      "end": {
                        "col": 19,
                        "ln": 20,
                      },
                      "start": {
                        "col": 14,
                        "ln": 20,
                      },
                    },
                    "type": "identifier",
                    "value": "bonus",
                  },
                },
              ],
            },
          },
        },
      },
    ],
  },
}
`;
